In [1]:
import numpy as np
import pandas as pd
data = pd.read_csv("C:/Users/Administrator/Downloads/Absenteeism_at_work_AAA/Absenteeism_at_work.csv", delimiter = ';')

In [2]:
data.head()

ID  Reason for absence  Month of absence  Day of the week  Seasons  \
0  11                  26                 7                3        1   
1  36                   0                 7                3        1   
2   3                  23                 7                4        1   
3   7                   7                 7                5        1   
4  11                  23                 7                5        1   

   Transportation expense  Distance from Residence to Work  Service time  Age  \
0                     289                               36            13   33   
1                     118                               13            18   50   
2                     179                               51            18   38   
3                     279                                5            14   39   
4                     289                               36            13   33   

   Work load Average/day             ...              Disciplinary failure  \
0                 239.554            ...                                 0   
1                 239.554            ...                                 1   
2                 239.554            ...                                 0   
3                 239.554            ...                                 0   
4                 239.554            ...                                 0   

   Education  Son  Social drinker  Social smoker  Pet  Weight  Height  \
0          1    2               1              0    1      90     172   
1          1    1               1              0    0      98     178   
2          1    0               1              0    0      89     170   
3          1    2               1              1    0      68     168   
4          1    2               1              0    1      90     172   

   Body mass index  Absenteeism time in hours  
0               30                          4  
1               31                          0  
2               31                          2  
3               24                          4  
4               30                          2  

[5 rows x 21 columns]

## Here output column is discrete and also having 19 distinct values. So, we are making 6 bins for hours in absenteeism(Output column) 

In [3]:
data['Absenteeism time in hours'] = data['Absenteeism time in hours'].astype(int)
lista = []
for i in data['Absenteeism time in hours'].tolist():
    if(i == 0):
        i = 'Zero'
    else:
        if(i<=5):
            i = '1-to-5'
        else:
            if(i<=16):
                i = '7-to-16'
            else:
                if(i<=40):
                    i = '24-to-40'
                else:
                    if(i<=80):
                        i = '40-to-80'
                    else:
                        if(i<=120):
                            i = '80-to-120'
    lista.append(i)    

In [4]:
data['absenteeism_cat'] = lista
data = data.drop('ID', axis=1)
data = data.drop('Absenteeism time in hours', axis = 1)

In [5]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import train_test_split

In [6]:
from sklearn.neighbors import KNeighborsClassifier

In [7]:
train, test = train_test_split(data, test_size = 0.3)

In [8]:
train_x = train.drop('absenteeism_cat', axis=1)
train_y = train['absenteeism_cat']
test_x = test.drop('absenteeism_cat', axis = 1)
test_y = test['absenteeism_cat']

In [9]:
m1 = DecisionTreeClassifier(max_depth=2)
m2 = RandomForestClassifier(n_estimators=5)
m3 = AdaBoostClassifier()
m4 = KNeighborsClassifier()

In [10]:
m1.fit(train_x, train_y)
m2.fit(train_x, train_y)
m3.fit(train_x, train_y)
m4.fit(train_x, train_y)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')

In [11]:
pred1 = m1.predict(test_x)
pred2 = m2.predict(test_x)
pred3 = m3.predict(test_x)
pred4 = m4.predict(test_x)

In [12]:
df1 = pd.DataFrame({'act':test_y, 'pred':pred1})
df2 = pd.DataFrame({'act':test_y, 'pred':pred2})
df3 = pd.DataFrame({'act':test_y, 'pred':pred3})
df4 = pd.DataFrame({'act':test_y, 'pred':pred4})

In [13]:
df1['stat'] = df1['act'] == df1['pred']
df2['stat'] = df2['act'] == df2['pred']
df3['stat'] = df3['act'] == df3['pred']
df4['stat'] = df4['act'] == df4['pred']

In [19]:
print(df1['stat'].value_counts()/ df1.shape[0] *100)
print(df2['stat'].value_counts()/ df2.shape[0] *100)
print(df3['stat'].value_counts()/ df3.shape[0] *100)
print(df4['stat'].value_counts()/ df4.shape[0] *100)

True     78.378378
False    21.621622
Name: stat, dtype: float64
True     72.522523
False    27.477477
Name: stat, dtype: float64
True     60.36036
False    39.63964
Name: stat, dtype: float64
True     67.117117
False    32.882883
Name: stat, dtype: float64


In [20]:
print("DecisionTree accuracy : ", (df1['stat'].value_counts()/ df1.shape[0] *100)[1])
print("RandonForest accuracy : ", (df2['stat'].value_counts()/ df2.shape[0] *100)[1])
print("AdaBoost accuracy : ", (df3['stat'].value_counts()/ df3.shape[0] *100)[1])
print("KNNeighbours accuracy : ", (df4['stat'].value_counts()/ df4.shape[0] *100)[1])

DecisionTree accuracy :  78.37837837837837
RandonForest accuracy :  72.52252252252252
AdaBoost accuracy :  60.36036036036037
KNNeighbours accuracy :  67.11711711711712


## Here, we are getting 78 % accuracy with Decision Tree and 72 % accuracy with RandomForest